### PR TITLE
VB-2326: add remove excluded dates

### PR DESCRIPTION
**Pull request by infojav**
* List 'Excluded dates' for a prison and allow them to be added and removed.
* Also:
  * Moved and renamed some files for improve structure
  * Used 'excluded' rather than 'exclusion' throughout for consistency
  * Moved implementation of `formatDate` to `utils.ts`

Still to do on this ticket:
* add validation for date passed when removing an exclude date
* check test coverage

### CODE REVIEW

**Review by tpmcgowan — COMMENTED** [review]
Comments added. Probably need to discuss the architecture of the application as it looks like there's some confusion about forms, GET/POST, APIs, etc.

Also need to update the API definitions because of the revised endpoints. Instructions here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attributes_for_form_submission

File: server/routes/prisons/exclusionDates/exclusionDatesController.ts [hunk 1]
```diff
@@ -1,12 +1,76 @@
...
+  'July',
+  'August',
+  'September',
+  'October',
+  'November',
+  'December',
+] as const
+
+const transformDate = (date: string): string => {
+  const splitedDate = date.split('-').map(d => Number(d))
```
Comment (psoleckimoj) [inline server/routes/prisons/exclusionDates/exclusionDatesController.ts:20]: You could use https://date-fns.org/docs/Getting-Started for this. There's a few standard date formatting functions we've created in one of the utils files I believe.

File: server/data/visitSchedulerApiClient.test.ts [hunk 2]
```diff
@@ -103,6 +103,36 @@ describe('visitSchedulerApiClient', () => {
     })
   })
 
+  describe('addExcludedDate', () => {
+    it('should return prison including the excluded date', async () => {
+      const prison = TestData.prison({ excludeDates: ['2023-12-25', '2023-12-26'] })
+
+      fakeVisitSchedulerApi
+        .put(`/admin/prisons/prison/${prison.code}/exclude-date/add`)
```
Comment (tpmcgowan) [inline server/data/visitSchedulerApiClient.test.ts:111]: Haven't tried, but I'm not sure if the contents of the request body (`2023-12-26`) actually getting tested here 🤔 ?

See: https://github.com/nock/nock#specifying-request-body - and an example at https://github.com/ministryofjustice/book-a-prison-visit-staff-ui/blob/8c6fc456455bbc89765ccf17749ba75e6cc8fc64/server/data/orchestrationApiClient.test.ts#L96

File: server/routes/prisons/exclusionDates/index.ts [hunk 3]
```diff
@@ -1,16 +1,19 @@
 import { RequestHandler, Router } from 'express'
-import { Services } from '../../../services'
+import { Services, SessionTemplateService } from '../../../services'
```
Comment (tpmcgowan) [inline server/routes/prisons/exclusionDates/index.ts:2]: This `SessionTemplateService` import isn't needed. It is why your build has failed the lint check on CircleCi: https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-visits-internal-admin-ui/346/workflows/8d3a7b09-1faa-4b1d-b931-54b3b1267743/jobs/1811?invite=true#step-107-9

File: server/routes/prisons/exclusionDates/exclusionDatesController.test.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+import { appWithAllRoutes, flashProvider } from '../../testutils/appSetup'
+import { createMockPrisonService, createMockSessionTemplateService } from '../../../services/testutils/mocks'
+import TestData from '../../testutils/testData'
+import { FlashErrorMessage } from '../../../@types/visits-admin'
+
+let app: Express
+let flashData: Record<string, string | FlashErrorMessage>
+
+const prisonService = createMockPrisonService()
+const sessionTemplateService = createMockSessionTemplateService()
```
Comment (tpmcgowan) [inline server/routes/prisons/exclusionDates/exclusionDatesController.test.ts:15]: No need to mock this as I don't think this controller uses the `SessionTemplateService`?

File: server/routes/prisons/exclusionDates/exclusionDatesController.test.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+  app = appWithAllRoutes({ services: { prisonService, sessionTemplateService } })
+})
+
+afterEach(() => {
+  jest.resetAllMocks()
+})
+
+describe('Show excluded dates', () => {
+  it('GET /prisons/{:prisonId}/exclusion-dates', () => {
```
Comment (tpmcgowan) [inline server/routes/prisons/exclusionDates/exclusionDatesController.test.ts:37]: Doesn't look like any of the tests for the exclusion dates listing pages (`GET /prisons/{prisonCode}/exclusion-dates`) actually check that the exclusion dates get rendered as a table listing?

Comment (infojav) [reply server/routes/prisons/exclusionDates/exclusionDatesController.test.ts:37]: It's checking if the page is rendered, I added more test, to check excluded Dates table and message if excludedDates array is empty.

File: server/utils/utils.ts [hunk 6]
```diff
@@ -21,3 +23,12 @@ export const initialiseName = (fullName?: string): string | null => {
   const array = fullName.split(' ')
   return `${array[0][0]}. ${array.reverse()[0]}`
 }
+
+export const formatDate = (dateToFormat: string, dateFormat = 'd MMMM yyyy'): string | null => {
```
Comment (tpmcgowan) [inline server/utils/utils.ts:27]: Not strictly part of this PR, but I've moved the implementation of this function from where the filter definitions are in `nunjucksSetup.ts` to here (and then called it in the filter's definition). This means the same function can be used in app code as `formatDate(...)` and in templates via `"date string" | formatDate`

File: server/routes/prisons/excludedDates/excludedDatesController.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  public validate(): ValidationChain[] {
+    return [
+      body('excludeDate.*').trim().toInt(),
+      body('excludeDate', 'Enter a valid date')
+        .customSanitizer(date => {
+          return `${date.year}-${date.month.toString().padStart(2, '0')}-${date.day.toString().padStart(2, '0')}`
+        })
+        .isDate({ format: 'YYYY-MM-DD', strictMode: true }),
```
Comment (tpmcgowan) [inline server/routes/prisons/excludedDates/excludedDatesController.ts:74]: Thinking I might put this block as a function somewhere so we can re-use on other `govukDateInput` fields...🤔

Comment (hutcheonb-moj) [reply server/routes/prisons/excludedDates/excludedDatesController.ts:74]: Worth doing then when we come back through to refactor some of this validation, think we've learnt a fair amount about express validator as we've been working through these features

**Review by hutcheonb-moj — APPROVED** [review]
LGTM
